### PR TITLE
[Enhancement] rank window function optimization, support row_number (3)

### LINKED ISSUE #5885: Rank window function optimization
## Enhancement

For the following sql

```sql
select * from (
    select *, rank() over (partition by v2 order by v3) as rk from t0
) sub_t0
where rk < 5;
```

The corresponding logical plan is as follows:

```
PREDICATE ( rank() < 5 )
            ↑
WINDOW_FUNCTION ( partition by v2 order by v3 )
            ↑
SORT ( order by v2, v3 )
            ↑
EXCHANGE ( shuffle by v2 )
            ↑
SCAN
```

**There are two main flaws of this plan:**

1. `EXCHANGE` Node needs to shuffle full data
2. `SORT` Node perform a machine level full sort

**Notice that there exists the predicate `rk < 5`, we can add a `PARTITION_TOPN` node before exchange node in order to filter data, the expected optimized plan is as follows:**

```
PREDICATE ( rank() < 5 )
            ↑
WINDOW_FUNCTION ( partition by v2 order by v3 )
            ↑
SORT ( order by v2, v3 )
            ↑
EXCHANGE ( shuffle by v2 )
            ↑
PARTITION_TOPN ( partition by v2, sort by v3, limit by rank )
            ↑
SCAN
```

**It should be emphasized that the output of the `PARTITION_TOPN` is still remained unsorted, the purpose of this operator is to filter data**

## Tasks

- [x] Support component partitioner, #5886 
    * only support one partition expr right now
- [x] Support PartitionTopN, #6118 
- [x] Support `row_number`, #6119 
- [x] Support `rank`, #6120 
    * by supporting TopN limit by rank
- [ ] Support `dense_rank`
    * by supporting TopN limit by dense_rank
- [ ] Support multi partition exprs
- [ ] Optimize PartitionTopN by partitioner supporting streaming style
    * Currently, data will be sent to `topn_sorter` until all the data have been partitioned, which is saved in hash table. But it is unnecessary because `topn_sorter` can help filter data so that we don't need to hold all the data
- [ ] Merge normal version and nullable version of PartitionHashMap, such as `PartitionHashMapWithOneNumberKey` and `PartitionHashMapWithOneNullableNumberKey`

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5885

## Enhancement

```sql
select * from (
    select *, rank() over (partition by v2 order by v3) as rk from t0
) sub_t0
where rk < 5;
```

For rank window function, including `rank`、`dense_rank`、`row_number`, if it has a related predicate (`rk < 5`), then it can be optimized by inserting a `PartitionTopN` operator before the `Sort` operator of window function. 

**The main purpose of `PartitionTopN` is to filter data, and it's output still remain unordered. It consists of three components:**

* `partitioner`：Divide the input chunk based on the partition exprs
* `sorter(topn)`：Each partition has an instance of sorter and is sorted independently
* `gather`：fetch chunks from all sorters into one data stream, so the data is still unordered after gahtering. Moreover, gather is a only logical concetp, not an actual component

```


                                   ┌────► topn─────┐
                                   │               │
 (unordered)                       │               │                  (unordered)
 inputChunks ───────► partitioner ─┼────► topn ────┼─► gather ─────► outputChunks
                                   │               │
                                   │               │
                                   └────► topn ────┘
```

**The implementation on the optimizer and the executor is a little different:**

* In optimizer, for simplicity, we do not define a new pair of `{Logical/Physical}PartitionTopNOperator`  but reuse the existing `{Logical/Physical}TopNOperator` by adding a new field `partitionByExprs` to record the partition by information. Besides, we need to pay attentation to the following things: 
    * Make sure that we cannot derive sort property from PartitionTopN
    * Make sure that ExchangeNode not set limit if PartitionTopN
* In executor, we define a new pair of `LocalPartitionTopN{Sink/Source}Operator`, and we may use different implementation based on the field `partitionByExprs`
    * if `partitionByExprs` is unset or empty, then the original pair of `PartitionSortSinkOperator/LocalMergeSortSourceOperator` is used
    * if `partitionByExprs` is not empty, then pair of `LocalPartitionTopN{Sink/Source}Operator` is used

## Tasks

- [x] Support component partitioner
    * only support one partition expr right now
- [x] Support PartitionTopN
- [x] **Support `row_number`(this pr)**
- [ ] Support `rank`
    * by supporting TopN limit by rank
- [ ] Support `dense_rank`
    * by supporting TopN limit by dense_rank
- [ ] Support multi partition exprs

## Performance Improvement of this pr

**test Info**

* TPCDS-100g
* 3 be of which is 64c/128g

**test sql**
* the original window function of Q67 is `rank`, since we only support `row_number` right now, so I replace it with `row_number`

```sql
-- query 67
select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,row_number() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1200 and 1200+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
```

**test result**
|  before | after |
|:-- |:--|
| 20s | 11s |

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 132 / 143 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SortInfo.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|19|25|76.00%|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|17|19|89.47%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|13|14|92.86%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java|52|53|98.11%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/TopNImplementationRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ExchangeSortToMergeRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/planner/ExchangeNode.java|7|7|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/QueryStmt.java|1|1|100.00%|

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 127 / 143 (88.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SortInfo.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|19|25|76.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|17|19|89.47%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RankWindowFunctionTopNRule.java|49|53|92.45%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/TopNImplementationRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ExchangeSortToMergeRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/planner/ExchangeNode.java|7|7|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/QueryStmt.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+public class PushDownPredicateWindowRankRule extends TransformationRule {
+
+    public PushDownPredicateWindowRankRule() {
+        super(RuleType.TF_PUSH_DOWN_PREDICATE_WINDOW_RANK,
+                Pattern.create(OperatorType.LOGICAL_FILTER).
+                        addChildren(Pattern.create(OperatorType.LOGICAL_WINDOW, OperatorType.PATTERN_LEAF)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java:46]: Should check partition expr size less than 2 now, and can be remove the limit after we support multi-partition

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        long limit = rightChild.getBigint();
+
+        List<ColumnRefOperator> partitionByColumns = windowOperator.getPartitionExpressions().stream()
+                .map(ScalarOperator::<ColumnRefOperator>cast)
+                .collect(Collectors.toList());
+
+        // TODO(hcf) we will support multi-partition later
+        if (partitionByColumns.size() > 1) {
+            return Collections.emptyList();
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java:112]: Move to check

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/planner/ExchangeNode.java [hunk 3]
```diff
@@ -92,7 +93,16 @@ public ExchangeNode(PlanNodeId id, PlanNode inputNode, boolean copyConjuncts) {
...
-            limit = inputNode.limit;
+            if (inputNode instanceof SortNode) {
+                SortNode sortNode = (SortNode) inputNode;
+                if (CollectionUtils.isEmpty(sortNode.getSortInfo().getPartitionExprs())) {
+                    limit = inputNode.limit;
+                } else {
+                    unsetLimit();
+                }
+            } else {
+                limit = inputNode.limit;
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/planner/ExchangeNode.java:104]: If there is a limit demand pushed down, how to deal with it, can it be terminated in advance?

Comment (trueeyu) [reply fe/fe-core/src/main/java/com/starrocks/planner/ExchangeNode.java:104]:
```
select * from (     select *, rank() over (partition by v2 order by v3) as rk from t0 ) sub_t0 where rk < 5 limit 5;
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java [hunk 4]
```diff
@@ -149,5 +155,35 @@ public LogicalTopNOperator.Builder withOperator(LogicalTopNOperator topNOperator
...
+            this.orderByElements = orderByElements;
+            return this;
+        }
+
+        public LogicalTopNOperator.Builder setOffset(int offset) {
+            this.offset = offset;
+            return this;
+        }
+
+        public LogicalTopNOperator.Builder setLimit(int offset) {
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java:174]: limit?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java [hunk 5]
```diff
@@ -149,5 +155,35 @@ public LogicalTopNOperator.Builder withOperator(LogicalTopNOperator topNOperator
...
+            this.offset = offset;
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java:175]: this.limit?

Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java:175]: Low level mistake, I'll fix it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java (hunk 4 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java:174]: Low level mistake, I'll fix it

File: fe/fe-core/src/main/java/com/starrocks/planner/ExchangeNode.java (hunk 3 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/planner/ExchangeNode.java:104]: For the ordinary TopN, merge exchange will be performed at the receiver side, and the final size of the data is N (the limit value)

But for PartitionTopN, we cannot perform merge operation at the receiver side

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 127 / 141 (90.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SortInfo.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|19|23|82.61%|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|17|19|89.47%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java|49|53|92.45%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/TopNImplementationRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ExchangeSortToMergeRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/planner/ExchangeNode.java|7|7|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/QueryStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 127 / 141 (90.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SortInfo.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|19|23|82.61%|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|17|19|89.47%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java|49|53|92.45%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/TopNImplementationRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ExchangeSortToMergeRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/planner/ExchangeNode.java|7|7|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/QueryStmt.java|1|1|100.00%|

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Review by trueeyu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 144 / 161 (89.44%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SortInfo.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/QueryTransformer.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|20|24|83.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|23|27|85.19%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateWindowRankRule.java|52|56|92.86%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Operator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/TopNImplementationRule.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/SortProperty.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|12|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ExchangeSortToMergeRule.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/planner/ExchangeNode.java|7|7|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/analysis/QueryStmt.java|1|1|100.00%|

**Review by Seaven — APPROVED** [review]

**Pull request merged**